### PR TITLE
Create pre-commit hook to run golangci-lint

### PR DESCRIPTION
**Pull request by chatton**
## Description

closes: #3048

In order to configure the linter, everyone must run

```bash
make setup-pre-commit
```

This will create a symlink so that in the future we just need to modify the newly added script.

This script ensures we can't commit linted files that are out of sync with our CI. 

It will then only run the linter on any modified files.

### Commit Message / Changelog Entry

N/A Dev improvement

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: scripts/hooks/pre-commit.sh [hunk 1]
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
+
+# check_golangci_lint_version ensures that the local version of golangci-lint
+# matches the version specified in .github/workflows/golangci.yml. This makes sure
+# that if the pre-commit hook is run locally, the changes should align with CI.
+function check_golangci_lint_version(){
+  local git_root="$(git rev-parse --show-toplevel)"
+
+  # extract the version of golangci-lint from the CI workflow file.
+  local golang_lint_ci_version="$(grep ' version' ${git_root}/.github/workflows/golangci.yml | awk '{ print $NF }')"
```
Comment (chatton) [inline scripts/hooks/pre-commit.sh:10]: this is pretty hacky, definitely open to suggestions on how to improve this.

File: scripts/hooks/pre-commit.sh [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+  # sample output of "golangci-lint --version"
+  # golangci-lint has version 1.53.2 built with go1.20.4 from 59a7aaf on 2023-06-03T10:44:21Z
+  local local_golang_lint_version="$(golangci-lint --version | awk '{ print $4 }')"
```
Comment (DimitrisJim) [inline scripts/hooks/pre-commit.sh:14]: `golangci-lint version --format short` is also an option after some digging

**Comment by crodriguezvega** [comment]
Awesome, @chatton! Can you please also update the [development setup docs](https://github.com/cosmos/ibc-go/blob/main/docs/dev/development-setup.md)?

File: scripts/hooks/pre-commit.sh [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+# that if the pre-commit hook is run locally, the changes should align with CI.
+function check_golangci_lint_version(){
+  local git_root="$(git rev-parse --show-toplevel)"
+
+  # extract the version of golangci-lint from the CI workflow file.
+  local workflow_golangci_lint_version="$(grep ' version' ${git_root}/.github/workflows/golangci.yml | awk '{ print $NF }')"
+
+  local local_golangci_lint_version="$(golangci-lint version --format short)"
+
+  if [[ "${workflow_golangci_lint_version}" != "${local_golangci_lint_version}" ]]; then
```
Comment (DimitrisJim) [inline scripts/hooks/pre-commit.sh:14]: testing this locally! Bumped into something, workflow version has a `v` in front of it, the comparison here will always fail, you need more bash.

Comment (chatton) [reply scripts/hooks/pre-commit.sh:14]: so looks like if you install with go get, it includes the v, but not with other versions 🤔 

```bash
~/go/bin/golangci-lint version --format short
v1.53.1
```

**Review by DimitrisJim — APPROVED** [review]
I've been using locally for ~4 hours now and everything seems to be working fine for me so my first Tested ACK.

File: scripts/hooks/pre-commit.sh [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    echo "aborting pre-commit hook"
+    exit 1
+  fi
+}
+
+# run_hook formats all modified go files that have been staged and adds them to the commit.
+function run_hook() {
+  make lint-fix-changed
+  echo "formatting any staged go files"
+  go_files="$(git diff --staged --name-only | grep \.go$)"
```
Comment (chatton) [inline scripts/hooks/pre-commit.sh:28]: thanks @DimitrisJim for adding `--staged`, nice catch 💪

**Review by damiannolan — APPROVED** [review]
I haven't verified the awk and sed cmds but I trust you guys have it covered 😅 
LGTM

**Review by charleenfei — APPROVED** [review]
it works for me! thank you @chatton @DimitrisJim

**Pull request merged**